helix-core/src/selection.rs

```diff
@@ -727,22 +729,16 @@ pub fn keep_or_remove_matches(
 pub fn select_on_matches(
     text: RopeSlice,
     selection: &Selection,
-    regex: &crate::regex::Regex,
+    regex: &rope::Regex,
 ) -> Option<Selection> {
     let mut result = SmallVec::with_capacity(selection.len());
 
     for sel in selection {
-        // TODO: can't avoid occasional allocations since Regex can't operate on chunks yet
-        let fragment = sel.fragment(text);
-
-        let sel_start = sel.from();
-        let start_byte = text.char_to_byte(sel_start);
-
-        for mat in regex.find_iter(&fragment) {
+        for mat in regex.find_iter(text.regex_input_at(sel.from()..sel.to())) {
```

Comment:
I have this branch merged into my driver for testing and I found a reliable crash here:

<details><summary>Backtrace</summary>

```
thread 'main' panicked at 'invalid span 5292..5034', /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/regex-cursor-0.1.1/src/input.rs:632:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:578:5
   1: core::panicking::panic_fmt
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:67:14
   2: regex_cursor::input::Input<C>::set_span
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/regex-cursor-0.1.1/src/input.rs:632:9
   3: regex_cursor::input::Input<C>::set_start
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/regex-cursor-0.1.1/src/input.rs:549:9
   4: regex_cursor::util::iter::Searcher<C>::try_advance
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/regex-cursor-0.1.1/src/util/iter.rs:437:9
   5: regex_cursor::util::iter::Searcher<C>::advance
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/regex-cursor-0.1.1/src/util/iter.rs:382:15
   6: <regex_cursor::engines::meta::regex::FindMatches<C> as core::iter::traits::iterator::Iterator>::next
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/regex-cursor-0.1.1/src/engines/meta/regex.rs:1988:9
   7: helix_core::selection::select_on_matches
             at /nix/persist/home/michael/src/helix/hx/helix-core/src/selection.rs:737:20
   8: helix_term::commands::select_regex::{{closure}}
             at /nix/persist/home/michael/src/helix/hx/helix-term/src/commands.rs:1829:17
   9: helix_term::ui::regex_prompt::{{closure}}
             at /nix/persist/home/michael/src/helix/hx/helix-term/src/ui/mod.rs:72:36
  10: helix_term::ui::raw_regex_prompt::{{closure}}
             at /nix/persist/home/michael/src/helix/hx/helix-term/src/ui/mod.rs:130:29
  11: <alloc::boxed::Box<F,A> as core::ops::function::FnMut<Args>>::call_mut
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/alloc/src/boxed.rs:1980:9
  12: <helix_term::ui::prompt::Prompt as helix_term::compositor::Component>::handle_event
             at /nix/persist/home/michael/src/helix/hx/helix-term/src/ui/prompt.rs:649:17
  13: helix_term::compositor::Compositor::handle_event
             at /nix/persist/home/michael/src/helix/hx/helix-term/src/compositor.rs:152:19
  14: helix_term::application::Application::handle_terminal_events::{{closure}}
             at /nix/persist/home/michael/src/helix/hx/helix-term/src/application.rs:671:22
  15: helix_term::application::Application::event_loop_until_idle::{{closure}}
             at /nix/persist/home/michael/src/helix/hx/helix-term/src/application.rs:325:55
  16: helix_term::application::Application::event_loop::{{closure}}
             at /nix/persist/home/michael/src/helix/hx/helix-term/src/application.rs:299:57
  17: helix_term::application::Application::run::{{closure}}
             at /nix/persist/home/michael/src/helix/hx/helix-term/src/application.rs:1227:38
  18: hx::main_impl::{{closure}}
             at /nix/persist/home/michael/src/helix/hx/helix-term/src/main.rs:161:53
  19: tokio::runtime::park::CachedParkThread::block_on::{{closure}}
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/park.rs:282:63
  20: tokio::runtime::coop::with_budget
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/coop.rs:107:5
  21: tokio::runtime::coop::budget
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/coop.rs:73:5
  22: tokio::runtime::park::CachedParkThread::block_on
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/park.rs:282:31
  23: tokio::runtime::context::blocking::BlockingRegionGuard::block_on
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/context/blocking.rs:66:9
  24: tokio::runtime::scheduler::multi_thread::MultiThread::block_on::{{closure}}
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/scheduler/multi_thread/mod.rs:87:13
  25: tokio::runtime::context::runtime::enter_runtime
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/context/runtime.rs:65:16
  26: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/scheduler/multi_thread/mod.rs:86:9
  27: tokio::runtime::runtime::Runtime::block_on
             at /home/michael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/runtime.rs:350:45
  28: hx::main_impl
             at /nix/persist/home/michael/src/helix/hx/helix-term/src/main.rs:163:5
  29: hx::main
             at /nix/persist/home/michael/src/helix/hx/helix-term/src/main.rs:37:21
  30: core::ops::function::FnOnce::call_once
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

</details>

Reproduction steps:

* Select these five lines: https://github.com/rabbitmq/ra/blob/6a6dcda610108bcf34f2803b8d0b7099956d7a97/test/ra_checkpoint_SUITE.erl#L155-L159
* Use `sch` (subselecting `checkpoint` but the crash occurs on the `h` input)

Reply:
Interesting thanks for reporting! I will look into that later today

Reply:
I am also using Mike's driver branch(mine got really far behind master and I was too lazy to update it) and have been seeing some crashes as well, I'll try and get them to you later today with a reproduction

Reply:
thanks that helped me find a gap in my testing for the prefilter. I did more testing and found two more bugs there. The prefilter code really is very subtle and full of edgecases. Almost all bugs come from there. I published a new fixed version and updated the version used here.